src/analysis/normed_space/pi_Lp.lean

```diff
@@ -106,36 +138,120 @@ explaining why having definitionally the right uniformity is often important.
 
 variables [Π i, pseudo_metric_space (α i)] [Π i, pseudo_emetric_space (β i)] [fintype ι]
 
-include fact_one_le_p
+/-- Endowing the space `pi_Lp p β` with the `L^p` edistance. We register this instance
+separate from `pi_Lp.pseudo_emetric` since the latter requires the type class hypothesis
+`[fact (1 ≤ p)]` in order to prove the triangle inequality. Registering this separately allows
+for a future quasi-emetric structure on `pi_Lp p β` for `p < 1`. -/
+instance : has_edist (pi_Lp p β) :=
+{ edist := λ f g, if hp : p = 0 then by subst hp; exact {i | f i ≠ g i}.to_finite.to_finset.card
+    else (if p = ∞ then ⨆ i, edist (f i) (g i)
+    else (∑ i, (edist (f i) (g i) ^ p.to_real)) ^ (1/p.to_real)) }
+
+variable {β}
+lemma edist_eq_card (f g : pi_Lp 0 β) : edist f g = {i | f i ≠ g i}.to_finite.to_finset.card :=
+dif_pos rfl
+
+lemma edist_eq_sum {p : ℝ≥0∞} (hp : 0 < p.to_real) (f g : pi_Lp p β) :
+  edist f g = (∑ i, edist (f i) (g i) ^ p.to_real) ^ (1/p.to_real) :=
+begin
+  dsimp [edist],
+  rw ennreal.to_real_pos_iff at hp,
+  rw [dif_neg hp.1.ne', if_neg hp.2.ne],
+end
+
+lemma edist_eq_supr (f g : pi_Lp ∞ β) : edist f g = ⨆ i, edist (f i) (g i) :=
+begin
+  dsimp [edist],
+  rw [dif_neg ennreal.top_ne_zero, if_pos rfl]
+end
 
-private lemma pos : 0 < p := zero_lt_one.trans_le fact_one_le_p.out
+lemma edist_self (f : pi_Lp p β) : edist f f = 0 :=
```

Comment:
```suggestion
protected lemma edist_self (f : pi_Lp p β) : edist f f = 0 :=
```
or (preferrably) inline it.

Reply:
I'll protect it. It's not inlined because this doesn't require the hypothesis that `[fact (1 ≤ p)]`, and these will be (eventually) used for the quasi-emetric structure for `p < 1`.

Reply:
Wait, what does"quasi-metric" mean? Surely it's not https://encyclopediaofmath.org/wiki/Quasi-metric ? Is it `pseudo_emetric_space`? In that case you can inline this lemma within the `pseudo_emetric_space` instance.

Reply:
No, it's this: https://en.wikipedia.org/wiki/Metric_(mathematics)#Semimetrics

This is one of those areas where the naming conventions are all over the place. I've always heard these referred to as quasi-metrics, and I've frequently heard people use the word semimetric for what we call a pseudometric in mathlib.

Reply:
That's annoying 😞

Reply:
I have phrased this as "relaxed triangle inequality" and mentioned that the literature varies; I mention both quasi-metric and semi-metric as names which appear.